pytest_playwright/pytest_playwright.py

```diff
@@ -145,22 +181,89 @@ def browser(launch_browser: Callable[[], Browser]) -> Generator[Browser, None, N
     browser = launch_browser()
     yield browser
     browser.close()
+    artifacts_folder.cleanup()
 
 
 @pytest.fixture
 def context(
-    browser: Browser, browser_context_args: Dict
+    browser: Browser,
+    browser_context_args: Dict,
+    pytestconfig: Any,
+    request: pytest.FixtureRequest,
 ) -> Generator[BrowserContext, None, None]:
+    pages: List[Page] = []
     context = browser.new_context(**browser_context_args)
+    context.on("page", lambda page: pages.append(page))
+
+    tracing_option = pytestconfig.getoption("--tracing")
+    capture_trace = tracing_option in ["on", "retain-on-failure"]
+    if capture_trace:
+        context.tracing.start(
+            name=slugify(request.node.nodeid),
+            screenshots=True,
+            snapshots=True,
+        )
+
     yield context
+
+    if capture_trace:
+        retain_trace = tracing_option == "on" or (
+            request.node.rep_call.failed and tracing_option == "retain-on-failure"
+        )
+        if retain_trace:
+            trace_path = _build_artifact_test_folder(pytestconfig, request, "trace.zip")
+            context.tracing.stop(path=trace_path)
+        else:
+            context.tracing.stop()
+
+    screenshot_option = pytestconfig.getoption("--screenshot")
+    capture_screenshot = screenshot_option == "on" or (
+        request.node.rep_call.failed and screenshot_option == "only-on-failure"
+    )
+    if capture_screenshot:
+        for index, page in enumerate(pages):
+            human_readable_status = (
+                "failed" if request.node.rep_call.failed else "finished"
+            )
+            screenshot_path = _build_artifact_test_folder(
+                pytestconfig, request, f"test-{human_readable_status}-{index+1}.png"
+            )
+            try:
+                page.screenshot(timeout=5000, path=screenshot_path)
+            except Error:
+                pass
+
     context.close()
 
+    video_option = pytestconfig.getoption("--video")
+    capture_video = video_option == "on" or (
+        request.node.rep_call.failed and video_option == "only-on-failure"
+    )
+    preserve_video = (
+        capture_video
+        and video_option == "on"
+        or (request.node.rep_call.failed and video_option == "retain-on-failure")
+    )
+    if preserve_video:
+        for page in pages:
+            video = page.video
+            if not video:
+                continue
+            try:
+                video_path = video.path()
+                file_name = os.path.basename(video_path)
+                video.save_as(
```

Comment:
Looking at this code, it seems like we only support sync api in our pytest fixtures?  I had no idea...

Reply:
Yes we only support the sync variant since the majority of the users is using sync (more common in Python + Pytest).